File: app/src/main/java/com/ctyeung/linearregression/MyPaperView.java
package com.ctyeung.linearregression;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

/*
 * view to render points and lines
 * - listens to touch actions (callback)
 */
public class MyPaperView extends View
{
    private final int regressionColor = Color.BLACK;
    private final int dotColor = Color.BLUE;
    private final int tangentColor = Color.GREEN;
    // defines paint and canvas
    private Paint drawPaint;
    private Path path;
    protected List<MyPoint> points = null;
    private PaperEvent mListener;

    public int width;
    public int height;

    private Line mLineRegression;

    public void setRegressionLine(MyPoint p0,
                                  MyPoint p1)
    {
        this.mLineRegression = new Line(p0, p1);
        render();
    }

    public MyPaperView(Context context,
                       AttributeSet attrs) {
        super(context, attrs);
        setFocusable(true);
        setFocusableInTouchMode(true);
        path = new Path();
        points = new ArrayList<MyPoint>();
    }

    public void setListener(PaperEvent listener)
    {
        mListener = listener;
    }

    public List<MyPoint> getPoints()
    {
        return points;
    }

    @Override
    protected void onDraw(Canvas canvas)
    {
        drawRegressionLine(canvas);
        drawTouchPoints(canvas);
        drawTangentLines(canvas);
    }

    /*
     * Draw regression line
     */
    private void drawRegressionLine(Canvas canvas)
    {
        drawPaint = new Paint();
        drawPaint.setAntiAlias(true);
        drawPaint.setStrokeWidth(5);
        drawPaint.setStyle(Paint.Style.STROKE);
        drawPaint.setStrokeJoin(Paint.Join.ROUND);
        drawPaint.setStrokeCap(Paint.Cap.ROUND);
        drawPaint.setColor(regressionColor);
        canvas.drawPath(path, drawPaint);
    }

    /*
     * Draw touch points
     */
    private void drawTouchPoints(Canvas canvas)
    {
        Paint dotPaint = new Paint();
        dotPaint.setStyle(Paint.Style.FILL);
        dotPaint.setColor(dotColor);

        for (MyPoint p : points) {
            // highlight point
            canvas.drawCircle((float) p.x, (float) p.y, 5, dotPaint);
        }
    }

    /*
     * Draw tangent lines
     */
    private void drawTangentLines(Canvas canvas)
    {
        drawPaint = new Paint();
        drawPaint.setAntiAlias(true);
        drawPaint.setStrokeWidth(3);
        drawPaint.setStyle(Paint.Style.STROKE);
        drawPaint.setStrokeJoin(Paint.Join.ROUND);
        drawPaint.setStrokeCap(Paint.Cap.ROUND);
        drawPaint.setColor(tangentColor);

        for (MyPoint p : points) {

            Line line = mLineRegression.findNormalLineFrom(p);
            MyPoint pp = mLineRegression.findIntersectionFrom(line);

            if(null!= pp) {
                Path tpath = new Path();
                tpath.moveTo((float) p.x, (float) p.y);
                tpath.lineTo((float) pp.x, (float) pp.y);
                canvas.drawPath(tpath, drawPaint);
            }
        }
    }

    @Override
    protected void onMeasure(final int widthMeasureSpec, final int heightMeasureSpec) {
        height= MeasureSpec.getSize(heightMeasureSpec);
        width= MeasureSpec.getSize(widthMeasureSpec);
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    protected boolean render()
    {
        path = new Path();
        if (null==points || points.size()<2)
            return false;

        if(points.size() > 2)
        {
            // render regression line
            path.moveTo((float)mLineRegression.p0.x, (float)mLineRegression.p0.y);
            path.lineTo((float)mLineRegression.p1.x, (float)mLineRegression.p1.y);
        }

        postInvalidate(); // Indicate view should be redrawn
        return true;
    }

    public void clear()
    {
        points.clear();
        path = new Path();
        postInvalidate(); // Indicate view should be redrawn
    }

    @Override
    public boolean onTouchEvent(MotionEvent event)
    {
        MyPoint point = new MyPoint(event.getX(), event.getY());

        // Checks for the event that occurs
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                // Starts a new line in the path
                break;

            /*case MotionEvent.ACTION_MOVE:
                path.lineTo(point.x, point.y);
                break;*/

            case MotionEvent.ACTION_UP:
                points.add(point);
                mListener.onActionUp();
                break;

            default:
                return false;
        }

        return true;
    }
}

File: xml_layout/LinearRegression/app/src/main/java/com/ctyeung/linearregression/views/MyPaperView.kt
package com.ctyeung.linearregression.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PointF
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

/*
 * view to render points and lines
 * - listens to touch actions (callback)
 */
class MyPaperView(  context: Context,
                    attrs: AttributeSet) :
    View(context, attrs) {

    val regressionColor = Color.BLACK
    val dotColor = Color.BLUE
    val tangentColor = Color.GREEN

    // defines paint and canvas
    var drawPaint: Paint? = null
    var path: Path? = null
    var listPoints: MutableList<PointF>? = null
    var mListener: PaperEvent? = null

    private var mLineRegression: Line? = null

    fun setRegressionLine(
        p0: PointF,
        p1: PointF
    ) {
        this.mLineRegression = Line(p0, p1)
        render()
    }

    init {
        isFocusable = true
        isFocusableInTouchMode = true
        path = Path()
        listPoints = ArrayList()
    }

    fun setListener(listener: PaperEvent) {
        mListener = listener
    }

    fun getPoints(): List<PointF> {
        val points:List<PointF> = listPoints!!.sortedWith(Comparators.CompareX)
        return points;
    }

    override fun onDraw(canvas: Canvas) {
        drawRegressionLine(canvas)
        drawTouchPoints(canvas)
        drawTangentLines(canvas)
    }

    /*
     * Draw regression line
     */
    private fun drawRegressionLine(canvas: Canvas) {
        drawPaint = Paint()
        drawPaint!!.isAntiAlias = true
        drawPaint!!.strokeWidth = 5f
        drawPaint!!.style = Paint.Style.STROKE
        drawPaint!!.strokeJoin = Paint.Join.ROUND
        drawPaint!!.strokeCap = Paint.Cap.ROUND
        drawPaint!!.color = regressionColor
        canvas.drawPath(path!!, drawPaint!!)
    }

    /*
     * Draw touch points
     */
    private fun drawTouchPoints(canvas: Canvas) {
        val dotPaint = Paint()
        dotPaint.style = Paint.Style.FILL
        dotPaint.color = dotColor

        for (p in listPoints!!) {
            // highlight point
            canvas.drawCircle(p.x.toFloat(), p.y.toFloat(), 5f, dotPaint)
        }
    }

    /*
     * Draw tangent lines
     */
    private fun drawTangentLines(canvas: Canvas) {
        drawPaint = Paint()
        drawPaint!!.isAntiAlias = true
        drawPaint!!.strokeWidth = 3f
        drawPaint!!.style = Paint.Style.STROKE
        drawPaint!!.strokeJoin = Paint.Join.ROUND
        drawPaint!!.strokeCap = Paint.Cap.ROUND
        drawPaint!!.color = tangentColor

        for (p in listPoints!!) {

            val line = mLineRegression!!.findNormalLineFrom(p)
            val pp = mLineRegression!!.findIntersectionFrom(line)

            if (null != pp) {
                val tpath = Path()
                tpath.moveTo(p.x.toFloat(), p.y.toFloat())
                tpath.lineTo(pp.x.toFloat(), pp.y.toFloat())
                canvas.drawPath(tpath, drawPaint!!)
            }
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        //height = View.MeasureSpec.getSize(heightMeasureSpec)
        //width = View.MeasureSpec.getSize(widthMeasureSpec)
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    protected fun render(): Boolean {
        path = Path()
        if (null == listPoints || listPoints!!.size < 2)
            return false

        if (listPoints!!.size > 2) {
            // render regression line
            path!!.moveTo(mLineRegression!!.p0.x.toFloat(), mLineRegression!!.p0.y.toFloat())
            path!!.lineTo(mLineRegression!!.p1.x.toFloat(), mLineRegression!!.p1.y.toFloat())
        }

        postInvalidate() // Indicate view should be redrawn
        return true
    }

    fun clear() {
        listPoints!!.clear()
        path = Path()
        postInvalidate() // Indicate view should be redrawn
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val point = PointF(event.x, event.y)

        // Checks for the event that occurs
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
            }

            /*case MotionEvent.ACTION_MOVE:
                path.lineTo(point.x, point.y);
                break;*/

            MotionEvent.ACTION_UP -> {
                listPoints!!.add(point)
                mListener!!.onActionUp()
            }

            else -> return false
        }// Starts a new line in the path

        return true
    }
}
